### PR TITLE
first attempt

### PR DESCRIPTION
**Pull request by andriimikhaylovskiy**

### CODE REVIEW

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 1]
```diff
@@ -30,7 +31,12 @@ public List<Book> findAll() {
     }
 
     @Override
-    public List<BookDto> getAll() {
+    public List<BookDto> getAll(java.awt.print.Pageable pageable) {
```
Comment (koval-kn) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: not that Pageable
```suggestion
    public List<BookDto> getAll(Pageable pageable) {
```

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 2]
```diff
@@ -30,7 +31,12 @@ public List<Book> findAll() {
...
+        return null;
```
Comment (koval-kn) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: its not implemented

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 3]
```diff
@@ -30,7 +31,12 @@ public List<Book> findAll() {
...
+    }
+
+    @Override
+    public List<BookDto> getAll(Pageable pageable) {
         throw new UnsupportedOperationException("Method not implemented");
```
Comment (koval-kn) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: its not implemented

File: src/main/java/mate/academy/spring/boot/service/BookService.java [hunk 4]
```diff
@@ -1,5 +1,6 @@
 package mate.academy.spring.boot.service;
 
+import java.awt.print.Pageable;
```
Comment (koval-kn) [inline src/main/java/mate/academy/spring/boot/service/BookService.java]: invalid import

**Review by koval-kn — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/spring/boot/service/BookService.java [hunk 5]
```diff
@@ -11,7 +12,9 @@ public interface BookService {
 
     List<Book> findAll();
 
-    List<BookDto> getAll();
+    List<BookDto> getAll(Pageable pageable);
+
+    List<BookDto> getAll(org.springframework.data.domain.Pageable pageable);
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/spring/boot/service/BookService.java]:
```suggestion
```
you have findAll for pagination

File: src/main/java/mate/academy/spring/boot/service/BookService.java [hunk 6]
```diff
@@ -11,7 +12,9 @@ public interface BookService {
 
     List<Book> findAll();
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/spring/boot/service/BookService.java]:
```suggestion
    List<Book> findAll(Pageable pageable);
```

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 7]
```diff
@@ -30,8 +31,13 @@ public List<Book> findAll() {
...
-    public List<BookDto> getAll() {
-        throw new UnsupportedOperationException("Method not implemented");
+    public List<BookDto> getAll(java.awt.print.Pageable pageable) {
+        return null;
+    }
+
+    @Override
+    public List<BookDto> getAll(Pageable pageable) {
+        return null;
     }
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: delete it, you have findAll() for this logic

**Review by MAMentorFX — CHANGES_REQUESTED** [review]
Use org.springframework.data.domain Pageable only. Implement Swagger documentation for all controller's methods. Dont create getAll, you have findAll.

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 8]
```diff
@@ -17,38 +19,35 @@
...
 
-    @Autowired
-    public BookController(BookService bookService) {
-        this.bookService = bookService;
-    }
-
-    @GetMapping
-    public List<BookDto> getAll() {
-        return bookService.getAll();
-    }
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]: you still need endpoint to get all Books

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 9]
```diff
@@ -25,13 +33,10 @@ public Book save(Book book) {
...
-        return bookRepository.findAll();
-    }
-
-    @Override
-    public List<BookDto> getAll() {
-        throw new UnsupportedOperationException("Method not implemented");
+    public List<BookDto> findAll(Pageable pageable) {
+        return bookRepository.findAll(pageable).stream()
+                .map(bookMapper::toDto)
+                .toList();
```
Comment (MAMentorFX) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: I think MapStruct supports mapping Page<> to List<>. like:
`List<BookDto> toDtos(Page<Book> pageable)`

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/spring/boot/controller/BookController.java (hunk 8 above)
Comment (sarakhmen) [reply src/main/java/mate/academy/spring/boot/controller/BookController.java]: please pay attention to the previous mentor's comment

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 10]
```diff
@@ -17,39 +20,39 @@
...
 @RestController
-@RequestMapping("/books")
+@RequestMapping(value = "/api/books")
 public class BookController {
     private final BookService bookService;
 
-    @Autowired
-    public BookController(BookService bookService) {
-        this.bookService = bookService;
-    }
```
Comment (sarakhmen) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]: why did you decide to remove the constructor injection? It's more preferable over field injection

File: src/main/java/mate/academy/spring/boot/repository/BookRepository.java [hunk 11]
```diff
@@ -1,8 +1,11 @@
 package mate.academy.spring.boot.repository;
 
 import mate.academy.spring.boot.model.Book;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
 public interface BookRepository extends JpaRepository<Book, Long>, JpaSpecificationExecutor<Book> {
+    Page<Book> findAll(Pageable pageable);
```
Comment (sarakhmen) [inline src/main/java/mate/academy/spring/boot/repository/BookRepository.java]: do you really need this method? Doesn't JpaRepository provide you with this method by default?

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Let's slightly improve your solution :)

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 12]
```diff
@@ -17,8 +20,9 @@
 import org.springframework.web.bind.annotation.ResponseStatus;
 import org.springframework.web.bind.annotation.RestController;
 
+@Tag(name = "Book management", description = "Endpoints for managing books")
 @RestController
-@RequestMapping("/books")
+@RequestMapping(value = "/api/books")
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]: why did you change it? /api prefix should better be set in .properties

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java (hunk 3 above)
Comment (okuzan) [reply src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: not fixed

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java (hunk 9 above)
Comment (okuzan) [reply src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: have you tried it?

File: src/main/java/mate/academy/spring/boot/model/Book.java [hunk 13]
```diff
@@ -5,17 +5,21 @@
...
 import java.math.BigDecimal;
 import lombok.Getter;
 import lombok.Setter;
+import lombok.ToString;
 import org.hibernate.annotations.SQLDelete;
 import org.hibernate.annotations.SQLRestriction;
 
 @Getter
 @Setter
+@ToString
```
Comment (okuzan) [inline src/main/java/mate/academy/spring/boot/model/Book.java]: it's ok for now, but be careful with related entities, can cause Lazy Init exception

**Review by okuzan — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 14]
```diff
@@ -48,35 +54,72 @@
...
-			<groupId>org.springframework</groupId>
-			<artifactId>spring-web</artifactId>
-			<version>6.1.4</version>
+			<version>1.5.3.Final</version>
 		</dependency>
 		<dependency>
 			<groupId>org.liquibase</groupId>
 			<artifactId>liquibase-core</artifactId>
-			<version>${liquibase.version}</version>
+			<version>4.27.0</version>
```
Comment (Elena-Bruyako) [inline pom.xml:62]: spring-parent already provides the liquibase version

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 15]
```diff
@@ -27,29 +31,43 @@ public BookController(BookService bookService) {
...
+            summary = "Search for books",
+            description = "Retrieve a list of books based on the provided search parameters."
+    )
+    public List<BookDto> searchBooks(BookSearchParametersDto searchParameters, Pageable pageable) {
+        return bookService.search(searchParameters, pageable);
+    }
+
+    @GetMapping("/books")
+    @Operation(summary = "Get all books", description = "Get the all book")
+    public List<BookDto> getAllBooks() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]: What the difference between getAllBooks and findAll? 
let's leave List<BookDto> getAll(@ParameterObject @PageableDefault Pageable pageable)

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 16]
```diff
@@ -27,29 +31,43 @@ public BookController(BookService bookService) {
...
     }
 
     @GetMapping("/search")
-    public List<BookDto> searchBooks(BookSearchParametersDto searchParameters) {
-        return bookService.search(searchParameters);
+    @Operation(
+            summary = "Search for books",
+            description = "Retrieve a list of books based on the provided search parameters."
+    )
+    public List<BookDto> searchBooks(BookSearchParametersDto searchParameters, Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]:
```suggestion
    public List<BookDto> searchBooks(BookSearchParametersDto searchParameters, @ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 17]
```diff
@@ -27,29 +31,43 @@ public BookController(BookService bookService) {
...
     @DeleteMapping("/{id}")
-    public void delete(@PathVariable Long id) {
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @Operation(summary = "Delete the book", description = "Delete the book by ID")
+    public void deleteBookById(@PathVariable Long id) {
         bookService.deleteById(id);
     }
 
     @GetMapping("/{id}")
+    @Operation(summary = "Get books", description = "Get the book by ID")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]:
```suggestion
    @Operation(summary = "Get book by id", description = "Get the book by ID")
```

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 18]
```diff
@@ -27,29 +31,43 @@ public BookController(BookService bookService) {
...
     public BookDto getBookById(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]:
```suggestion
    public BookDto getBookById(@PathVariable @Positive Long id) {
```
Don't forget to add `@Validated` under the controller

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 19]
```diff
@@ -27,29 +31,43 @@ public BookController(BookService bookService) {
...
-    public List<BookDto> getAll() {
-        return bookService.getAll();
-    }
-
-    @ResponseStatus(HttpStatus.NO_CONTENT)
     @DeleteMapping("/{id}")
-    public void delete(@PathVariable Long id) {
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @Operation(summary = "Delete the book", description = "Delete the book by ID")
+    public void deleteBookById(@PathVariable Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]:
```suggestion
    public void deleteBookById(@PathVariable @Positive Long id) {
```

File: src/main/java/mate/academy/spring/boot/mapper/BookMapper.java [hunk 20]
```diff
@@ -1,18 +1,19 @@
...
 public interface BookMapper {
-    BookMapper INSTANCE = Mappers.getMapper(BookMapper.class);
 
     @Mapping(target = "id", ignore = true)
     Book createBookRequestDtoToBook(CreateBookRequestDto dto);
 
-    BookDto bookToBookDto(Book book);
+    BookDto toDto(Book book);
+
+    Book toModel(CreateBookRequestDto requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/mapper/BookMapper.java:15]: Why do you need toModel and createBookRequestDtoToBook ?

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 21]
```diff
@@ -1,36 +1,48 @@
...
+import org.springframework.data.jpa.domain.Specification;
 import org.springframework.stereotype.Service;
 
 @Service
 public class BookServiceImpl implements BookService {
     private final BookRepository bookRepository;
+    private final BookMapper bookMapper;
+    private final BookSpecificationBuilder bookSpecificationBuilder;
 
     @Autowired
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: Don't need to use @Autowired if you have only one constructor. 
Use `@RequiredArgsConstructor`

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 22]
```diff
@@ -1,36 +1,48 @@
...
+        this.bookMapper = bookMapper;
+        this.bookSpecificationBuilder = bookSpecificationBuilder;
     }
 
     @Override
-    public Book save(Book book) {
-        return bookRepository.save(book);
+    public BookDto save(CreateBookRequestDto requestDto) {
+        Book book = bookMapper.toModel(requestDto);
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 23]
```diff
@@ -50,7 +62,11 @@ public void deleteById(Long id) {
...
 
     @Override
-    public List<BookDto> search(BookSearchParametersDto searchParameters) {
-        return Collections.emptyList();
+    public List<BookDto> search(BookSearchParametersDto params, Pageable pageable) {
+        Specification<Book> bookSpecification = bookSpecificationBuilder.build(params);
+        return bookRepository.findAll(bookSpecification, pageable)
+                .stream()
+                .map(bookMapper::toDto)
+                .toList();
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java:78]: Why don't you implement methods getBookById, createBook and deleteById?

File: src/main/java/mate/academy/spring/boot/specification/tools/book/search/parameters/BookSearchParameters.java [hunk 24]
```diff
@@ -0,0 +1,4 @@
+package mate.academy.spring.boot.specification.tools.book.search.parameters;
+
+public record BookSearchParameters(String[] titles, String[] authors) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/specification/tools/book/search/parameters/BookSearchParameters.java]:
```suggestion
public record BookSearchParameters(List<String> titles, List<String> authors) {
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/mate/academy/spring/boot/controller/BookController.java [hunk 25]
```diff
@@ -17,39 +24,52 @@
...
+    @Operation(summary = "Search for books",
+            description = "Retrieve a list of books based on the provided search parameters.")
+    public List<BookDto> searchBooks(BookSearchParametersDto searchParameters,
+                                     @ParameterObject @PageableDefault Pageable pageable) {
+        return bookService.search(searchParameters, pageable);
+    }
+
+    @GetMapping("/books")
+    @Operation(summary = "Get all books", description = "Get the all book")
+    public List<BookDto> getAll(@ParameterObject @PageableDefault Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/controller/BookController.java]: what the difference between getAll and findAll endpoints?

File: src/main/java/mate/academy/spring/boot/dto/BookDto.java [hunk 26]
```diff
@@ -11,6 +11,8 @@ public class BookDto {
     private String description;
     private String coverImage;
 
+    private boolean deleted;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/BookDto.java]: If it's response BookDto, stick to the task description 
Example of response body:
```
[
  {
    "id": 1,
    "title": "Sample Book 1",
    "author": "Author A",
    "isbn": "9781234567897",
    "price": 19.99,
    "description": "This is a sample book description.",
    "coverImage": "http://example.com/cover1.jpg"
  },
  {
    "id": 2,
    "title": "Sample Book 2",
    "author": "Author B",
    "isbn": "9789876543210",
    "price": 24.99,
    "description": "Another sample book description.",
    "coverImage": "http://example.com/cover2.jpg"
  }
]

```

Also, you can use @Data annotation to avoid generating get and set methods, or use java record

File: src/main/java/mate/academy/spring/boot/mapper/BookMapper.java [hunk 27]
```diff
@@ -1,18 +1,20 @@
...
 
-@Mapper
+@Mapper(config = MapperConfig.class)
 public interface BookMapper {
-    BookMapper INSTANCE = Mappers.getMapper(BookMapper.class);
 
     @Mapping(target = "id", ignore = true)
-    Book createBookRequestDtoToBook(CreateBookRequestDto dto);
+    @Mapping(target = "deleted", ignore = true)
+    // Book createBookRequestDtoToBook(CreateBookRequestDto dto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/mapper/BookMapper.java]: remove commented code

File: src/main/java/mate/academy/spring/boot/service/BookService.java [hunk 28]
```diff
@@ -4,20 +4,21 @@
...
 
-    List<Book> findAll();
+    List<BookDto> findAll(Pageable pageable);
 
-    List<BookDto> getAll();
+    List<BookDto> getAllBooks();
 
     BookDto getBookById(Long id);
 
     BookDto createBook(CreateBookRequestDto bookDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookService.java]: it looks like the same as BookDto save(CreateBookRequestDto requestDto);. Please, leave one

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java (hunk 3 above)
Comment (Elena-Bruyako) [reply src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: Remove pointless method

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 29]
```diff
@@ -1,42 +1,55 @@
...
     @Override
     public BookDto getBookById(Long id) {
-        throw new UnsupportedOperationException("Method not implemented");
+        return bookRepository.findById(id)
+                .map(bookMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Book with id: %s not found", id)));
     }
 
     @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java:58]: remove `public BookDto createBook(CreateBookRequestDto bookDto) {`

File: src/main/java/mate/academy/spring/boot/service/BookService.java [hunk 30]
```diff
@@ -4,20 +4,21 @@
 import mate.academy.spring.boot.dto.BookDto;
 import mate.academy.spring.boot.dto.BookSearchParametersDto;
 import mate.academy.spring.boot.dto.CreateBookRequestDto;
-import mate.academy.spring.boot.model.Book;
+import org.springframework.data.domain.Pageable;
 
 public interface BookService {
-    Book save(Book book);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookService.java:10]: Please, add update method and implement it

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 31]
```diff
@@ -50,7 +63,19 @@ public void deleteById(Long id) {
...
-        return Collections.emptyList();
+    public List<BookDto> search(BookSearchParametersDto params, Pageable pageable) {
+        Specification<Book> bookSpecification = bookSpecificationBuilder.build(params);
+        return bookRepository.findAll(bookSpecification, pageable)
+                .stream()
+                .map(bookMapper::toDto)
+                .toList();
+    }
+
+    private String generateUniqueIsbn() {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java:81]: You can provide validation for isbn by using pattern and validator. Don't need to generate it.

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Pay attention to the previous comments. Don't commit methods that are not implemented or pointless.

File: src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java [hunk 32]
```diff
@@ -1,56 +1,96 @@
...
+        return bookRepository.findAll(bookSpecification, pageable)
+                .stream()
+                .map(bookMapper::toDto)
+                .toList();
     }
 
     @Override
-    public List<BookDto> search(BookSearchParametersDto searchParameters) {
-        return Collections.emptyList();
+    public List<BookDto> getAllBooks(Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/service/BookServiceImpl.java]: You already have getAll method, remove

File: src/main/java/mate/academy/spring/boot/dto/BookDto.java [hunk 33]
```diff
@@ -2,68 +2,14 @@
...
-    }
+public record BookDto(
+        Long id,
+        String title,
+        String author,
+        String isbn,
+        BigDecimal price,
+        String description,
+        String coverImage,
+        boolean deleted
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/spring/boot/dto/BookDto.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's make solution perfect!

**Review by sarakhmen — APPROVED** [review]
good job :)
